### ISSUE TITLE
Send task to queue without executing it locally

### ISSUE THREAD

**Issue by H4dr1en** [allegroai/trains#128]
I am confused with `train_agent`s: they are supposed to pull experiments to run from queues, but I could not find in the documentation how to add tasks in the queue.

Considering the following setup:
- remote trains-server
- remote trains-agent (in same machine as trains-server for now, for debugging purpose)
- local computer

Let's consider the 2-liners example from the doc:
```
from trains import Task
task = Task.init(project_name="my project", task_name="my task")
```

If I run this code locally, this correctly registers the task (visible in the web UI), but how is this experiment supposed to be sent to the queue? This is only executed locally so far, so if I run

```
from trains import Task
task = Task.init(project_name="my project", task_name="my task")
# training code here
```

This will be executed in my local machine. How to send the task to the queue without running it locally, so that it is executed by the agent?

**Comment by bmartinn** [comment]
Hi @H4dr1en ,
Here is the missing part ;)
"right click" on the experiment in the UI, then select "clone".
On the cloned experiment, "right click" and select "enqueue", then select execution queue from the drop-down menu.

Please note that the newly cloned experiment is editable, all the fields can be edited.
This includes "base docker image", "installed packages" (you can think of it as the equivalent to the `requirements.txt`), "uncommitted changes" etc.

Probably the most important part is the "hyper-parameters" section, any change there will actually change the argparser or the connected dictionary ([example](https://github.com/allegroai/trains/blob/master/examples/hyper_parameters_example.py#L36)) values when the *trains-agent* is running the experiment.

Obviously you can do the same from code (i.e. clone experiment and send for remote execution), see automation [example](https://github.com/allegroai/trains/blob/master/examples/automl/task_piping_example.py)

**Comment by H4dr1en** [comment]
Thanks for the answer, so far I was going in the UI, canceled and restarted the experiment so that it would be enqueued.

> Notice the fact that in order to have the experiment show up in the UI in the first place, someone somewhere had to have run it manually (i.e. without trains-agent), from that point onward, you can just clone & enqueue :)

This is what confused me: what is the use-case of that behavior? It is a bit cumbersome to run it locally, kill the task, go the UI, clone failed experiment and restart it to properly enqueue it. It also creates a failed entry in UI each time I want to create a new experiment.

I had a look at the `pytorch_mnist.py` example, using a CLI solves the problem of not running the code locally, however how can we pass parameters in the UI? By editing the `SCRIPT PATH` entry of the experiment?

**Comment by bmartinn** [comment]
Hi @H4dr1en ,
The idea behind this workflow is that initial code unusually executed by someone when writing it in the first place (or debugging etc.) Then this execution automatically fills in all the entries in the Experiment.

That said you can do it manually: clone any experiment and change the git repo links / branch
initial script, working directory hyper-parameters etc.
The thing is, manually filling in all these details is almost as bad as writing a YAML file, which is what we want to avoid in the first place...


> Did you consider simply adding a trains register script.py --enqueue default command that only

What did you have in mind? 
Is the `script.py` a standalone, or in a git repo ?
I can diffidently see potential here, but could elaborate more on the use case you are thinking of?

**Comment by H4dr1en** [comment]
Thanks for the answer, this is much more clear now!

> What did you have in mind?

I don't think it is necessary to develop further since I had not fully understood the workflow at the time of writing. I will close this issue but give you a bit more details, in case you are interested in continuing this discussion:

What I had in mind was an easy way (command) to:
- create an experiment out of a script, for example `pytorch_mnist.py` (within a git repo)
- save this experiment in the server
- optional: add this experiment to queue

But having an argparse wrapper around the experiment logic defining hyper-parameters makes it easy to run locally and finetune/clone/edit/start later on.

**Comment by bmartinn** [comment]
Hi @H4dr1en , thank you for the feedback! 

My only thinking is if we have a script that does that, as you noted, in order to automatically grab the ArgParser, it needs to run the actual python code, and if we do that, then well, we can just add the trains import & init to the code itself ...

Obviously running the actual code (in your example  `pytorch_mnist.py` ),will automatically grab argparser and put it into hyper-parameters.
Then all you need is to sweep over the configuration, with something like this [one](https://github.com/allegroai/trains/blob/master/examples/automl/automl_random_search_example.py) that randomly samples parameters and puts a cloned copy of the experiment into an execution queue :)

**Comment by H4dr1en** [comment]
After giving a second though, I would like to reopen this issue because I have a valid use case that would justify having that feature:

As part of an automated pipeline, I would like to setup a github action that will, on each commit, create a Task an enqueue it. Here is the code that I would like to have:
```
def main(remote=False):

    # Note here that I would like to pass a `draft` parameter
    # and `init` would return a draft experiment
    draft_task = Task.init(project_name="test-trains", task_name="test-remote-run-singleton", reuse_last_task_id=True, draft=True)

    if distributed:  # We clone to get a draft task and enqueue it
        draft_task.set_parameter("distributed", False)
        Task.enqueue(draft_task)
        return
    else:
        # Do actual training

if __name__ == '__main__':
    main(remote=True)  # Use argparse so that Trains recognize it as a parameter
```

As you can see, the only thing preventing me to do that is that init returns a Task that as status `Running`, while I would like to have it as Draft. This would allows me to add it to queue or not, based on the "remote" parameter. With that I can either run it locally or remotely. WDYT?

PS: I also tried to 

```
    if distributed:  # We clone to get a draft task and enqueue it
        draft_task = Task.clone(task)    # Now I get a draft Task
        draft_task.set_parameter("distributed", False)
        Task.enqueue(draft_task)
        return
```
This works, but it will create two experiments in the UI, while I only want one. Also the first one (local) will succeed, but the second one will fail:

```
 from trains import Task
ModuleNotFoundError: No module named 'trains'
```
For some reason it did not add the trains dependency in the requirements (although existing in the requirements and imported in the script)

**Comment by bmartinn** [comment]
Hi @H4dr1en ,
Let me see if I understand the setup.
The idea is for GitHub actions to automatically launch (i.e send for execution) a task, this already sounds great :)
Now my question is what is the Task to be executed and what's the relation to the GitHub action.
Task contains a few sections actually describing the env setup itself (i.e. git repo installed packages etc.) , how would you fill those in ?
My thinking is "template task" (or a base task) that every time the action takes place, it actually clones the base_task, changes a few parameters and send for execution.
What am I missing?

**Comment by achigeor** [comment]
Jumping in on this (we work together with @H4dr1en). First of all kudos and hats off for the great work, really awesome, and thanks for your quick and elaborate responses - much appreciated!

## Context
So what we currently have _without trains_ is (some already covered here and in #130):
* We have an installable python package that has all the dependencies and different endpoints train / convert / test in the `setup.py` (#130)
* We connect to a training VM
* install with `-e .[train]`, `-e .[convert]`, `-e .[valid]` in different envs
* run sequentially: 
    * activate env [train/convert/valid]
    * run CLI command to [train/convert/valid]

Now, we want to leverage trains for:
* Experimentation tracking and reproducability
* Taking care of the workers and queues (done that) 

Ideally, one would run a `pipeline.py` script in any machine (let's focus on local, not github actions) that:
* Creates a _trains_ Task
* Runs that task directly on a worker **without** executing anything locally

Ideally, the above can happen in the following flow:
* You create a branch `fancy-new-experiment`
* Implement your new model / loss / training regime
* Commit your changes
* Run `pipeline.py` (let's focus on the user's local machine for this)
* Done (pipeline code is running on an automatically spinned worker)
* I can now check my experiments on the UI (and do as I please there)

On the topics that we discussed so far there seems to be two solutions, but both have problems as follows:

## Possible Solutions 
### _Possible Solution 1_ 
----
As @bmartinn you suggested:
* Create a Template task for every _type_ of experiment, let's say we have `Train Task Template`
* Run once, locally
* Go to UI, manually change the `Installed packages` to `.[train]`
* Every new training task is cloning the template task and enqueues it (happens in `piping.py` as in AutoML examples)

**Problem**: However, this is problematic because the template task has the commit hash bundled with it. This makes impossible to use the Template Task in different commits / different branches, since that any time you clone the template and enqueue it, the worker will checkout the commit hash defined in the template that you cloned from.
It also has the "problem" that any task template needs to be one of `training`, `testing` `inference`, since there's no support (that we know of) for custom task types.

### _Possible Solution 2_ 
----
As @H4dr1en wrote above (rather a hack):
```
from argparse import ArgumentParser
from trains import Task

def run(args):
    task = Task.init(project_name='foo', task_name='fancy-new-experiment', reuse_last_task_id=True))
    if args.distributed:  # We clone to get a draft task and enqueue it
        draft_task = Task.clone(task)    # Now I get a draft Task
        draft_task.set_parameter("distributed", False)
        Task.enqueue(draft_task)
        return
    print("we're on the worker")  # this will only be printed in the worker

if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument('-d', '--distributed', action='store_true',
                        help='Executes the task remotely using Trains, instead of locally.')
    args = parser.parse_args()
    run(args)
```
**Problem**: However, this is problematic because:
* We end up with 2 experiments in the UI (the local one that executed nothing and the one in the worker that actually runs), even thought we specify `reuse_last_task_id=True`
* We cannot specify the dependencies ourselves (as in the template case)  and will always use what is installed in the local machine at the moment of initialization

## Proposals
In general, allow for more flexibility on the creation of tasks. 

Specific **possible solution 1**:
Allow users to define the commit hash when cloning a task, thus overwritting the one inherited from the template.  One could then do:
```
template_task = Task.get_task(project_name='templates', task_name='Train Template')
my_current_task = Task.clone(source_task=template_task, project="fancy-new-project", name="fancy-new-experiment",
                             commit_hash={commit_hash}) # default is commit hash on template
Task.enqueue(my_current_task)
```
Not sure if thats currenly possible thought, at least we didn't find a way to do so.

For **possible solution 2**, but also as a general improvement (I like this the most):
* Allow users to do `Task.init` without running the task. This could probably be achieved by something like allowing the user to define a task as `Draft`, i.e. `task = Task.init(.., status="draft")`. Then you are can clone it and enqueue it without having to run twice
* Allow users to define a `requirements.txt` when creating a Task. Something like: `task = Task.init(..., dependencies='path/to/requirements.txt')`

----
WDYT? Would any of the two proposals be something that you see `trains` doing? Looking forward to your thoughts!

**Comment by bmartinn** [comment]
@achigeor thank you for the kind words, much appreciated !
I hope I come bearing good news :)

- When you clone a task, click on the "Execution" / "Source Code"  Edit button. Now you can change the "commit id" to "last commit in branch" and select the `fancy-new-experiment` branch.

- Let's think of this experiment as base "training" experiment. The `pipeline.py` will clone this experiment and send the cloned experiment to execution. The *trains-agent* will pull the experiment (latest in the branch at time of pulling), then fix the commit id on the cloned experiment. This give us the ability to trace back experiment on the one hand, and on the other have a template experiment that we can reuse.

- This means that we could change the `pipeline.py` by editing the template tasks, which is quite handy (think repo branch, python packages, base docker image etc.)

- Now the downside is that publishing (locking) an experiment you can only do on an executed experiment, so since the template task is in "draft mode" everyone can edit it, I guess that's also a bonus, depends how you think about it :) (also, well half true, you can always hack it from code and do Task.get_task(), task.started(), task.stopped(), task.publish())

- Regrading "Run pipeline.py (let's focus on the user's local machine for this)". In the upcoming version we will make it easier to create controllers. That means a small script (say `pipeline.py`) that only clones tasks, change parameters, wait for them to finish and setup the next step. The idea is first to make it easier so less boilerplate code, the second is to be able to enqueue `pipeline` Task as well, so I don't need to worry about my machine managing the pipeline. 

- This is what we dubbed "services mode". *trains-agent* in its current configuration can only execute one task at a time (which makes sense if we are thinking GPU/CPU hogging code), we now have another mode of the *trains-agent* --service-mode, where the agent will spin as many Tasks as you send it, but it's on the users to make sure these tasks are "service"/"monitor"/"control" tasks, and not cpu/ram intensive jobs. This allows us to put many pipeline tasks on a single machine, since most of the time they are sleeping. You can already start experimenting with `trains-agent==0.14.2rc2`

One last thought, obviously it is possible to "export"/"import" a task, at the end there is a REST API that you can access. But the question I'm raising is, what would be good practice. Our initial thinking was, okay let's just give users the ability to edit everything (which is still possible), but then you end up with yaml over yaml files to maintain, and your DLOps becomes messy and unreadable (because everything is in code, and you end up with YAML files in git etc...) This is why we designed Trains the way we did, so we could circumvent most of the manual configuration and more importantly maintenance. That said, if you feel strongly about manually editing Tasks, I'd be happy to point in the right direction.

If I missed any question / concern, let me know, I'll try to make it more clear.

BTW: If you do not follow the [slack channel](https://allegroai-trains.slack.com/), we just announced tags and custom Task will arrive in the next trains-server version, so I'm hoping this is good news on the automation part.

**Comment by H4dr1en** [comment]
Hi @bmartinn , Thanks for the detailed answer!

> When you clone a task, click on the "Execution" / "Source Code" Edit button. Now you can change the "commit id" to "last commit in branch" and select the fancy-new-experiment branch.

This is cool! But only solves part of the issue, because we have to clone based on a template that is already existing. For some use cases it can be enough, but for ours not.

> Regrading "Run pipeline.py (let's focus on the user's local machine for this)". In the upcoming version we will make it easier to create controllers. That means a small script (say pipeline.py) that only clones tasks, change parameters, wait for them to finish and setup the next step. The idea is first to make it easier so less boilerplate code, the second is to be able to enqueue pipeline Task as well, so I don't need to worry about my machine managing the pipeline.

Very cool, looking foward to that!

> why would you need to change the "Installed packages" in UI?
If you are running the code from the .train env, Trains should have caught it automatically, no?

Not really, it does not shows the local package installed with `.[train]`, therefore if we clone the experiment, it is not possible to reproduce the env (unless we add it back manually).

After some days of digging into the trains code, here is the approach we came up with:

```
def main(args):

    task = Task.current_task() if running_remotely() else Task.init()

    if args.remote and not running_remotely():

        task.reset()    # To convert to Draft
        # Trains correctly guess environment, except from local package itself
        # Therefore we overwrite requirements
        # This is only possible is sdk.development.vcs_repo_detect_async = false
        # Otherwise the overwriting happens before the vcs repo detection finishes.
        task._update_requirements([".[train]"])
        Task.enqueue(task)
        task.mark_stopped()  # Otherwise Task marked as finished
        logger.report_text(f"Successfully scheduled training Task (id={task._id})")
        return

    else:
        task.mark_started()
        # ACTUAL TRAINING


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Basic Parser')
    parser.add_argument('--remote', action='store_true', default=False,
                        help='Executes the task remotely using Trains, instead of locally.')
    parse_args = parser.parse_args()
    main(parse_args)
```

This allows us to have a `main.py` that we can use to start training locally using trains or remotely, also using trains (using `--remote` parameter).

**Comment by bmartinn** [comment]
Hi @H4dr1en 

> something fixed in the nightly version of trains-agent? We only used v0.14.1 so far.

Actually I think it was fixed (or at least improved) in [0.14.2](https://github.com/allegroai/trains/releases/tag/0.14.2) (see: *Improve repository and module detection*) Anyhow worth revisiting with 0.14.3

That said, good news, the next RC will contain `Task.add_requirements('package[.train]', version=None)` which will forcefully add a package into the "Installed Packages" list. Notice that if you do not specify a version it will try to look for the installed version, if it fails to find it, you will have the package appear as is. Then when the *trains-agent* will run the experiment, it will install everything, and replace with the specific versions that were used :)

Regrading the purposed solution, first kudos on the hack! really cool stuff :+1: 
I would also like to suggest an improvement on the purposed solution:

```python
    # if running remotely Task.init() does something similar to Task.current_task() so no need to worry about it.
    task = Task.init()

    if args.remote and not running_remotely():
        # since we are running "locally", first let's close the Task, 
        # it will update all the packages, analyze the code etc.
        task.close()
        task.reset()    # set status to Draft so we can Enqueue it.
        Task.enqueue(task)
        return  # Successfully scheduled remote task, exit.

    # ACTUAL TRAINING
```

Notice that `task.close()` will have 10 seconds timeout to detect the repository and all its packages, 
if this is too fast, just add `sleep()` before the `close()`. Because the analysis is done in the background, you just gave a it a few more seconds of grace, before shutting it down.

That actually connects (pun intended) to a few internal discussions we were having. 
1) Manually "importing" an experiment from a YAML file is always an option, even though I personally think that it beats the purpose of "automagically" detecting everything about the code.
That said we do contemplate on adding import/export functionality, and if you feel strongly about it, please open an issue. If we have enough people requesting it, it will get done :)
2) Doing exactly as you suggested, really this is exactly what we also came up with. 
Basically initialize the task, close it, make sure the analysis is completed (including argparser etc, ) and exiting the code.
The main issues we still need to solve (and any suggestions are welcome)
`-` close() should wait for the analysis instead of giving it fixed 10sec grace.
`-` argparser should somehow be called, otherwise, the argparser will not appear in the hyper-parameters section
`-` What happens if you do `Task.current_task.connect({'somewhere_in_mid_code': 'inside_a_few_nested', 'functions': 1})` . Since this code is never reached when creating the "template" we will not be able to add these parameters into the "hyper-parameters" section 

What do you think?

**Comment by H4dr1en** [comment]
Hi @bmartinn ,

> Actually I think it was fixed (or at least improved) in 0.14.2 (see: Improve repository and module detection) Anyhow worth revisiting with 0.14.3

Yes, in 0.14.3 the local packages are now correctly listed! 👍 

> That said, good news, the next RC will contain Task.add_requirements('package[.train]', version=None) 

Cool, looking forward to that! 👍  Thanks also for the improvement of the hack.

> Notice that task.close() will have 10 seconds timeout to detect the repository and all its packages, if this is too fast, just add sleep() before the close(). Because the analysis is done in the background, you just gave a it a few more seconds of grace, before shutting it down.

We simply set `sdk.development.vcs_repo_detect_async = false` to make sure that the detection is done sequentially. Setting a sleep is less robust I think.

> Manually "importing" an experiment from a YAML file is always an option, even though I personally think that it beats the purpose of "automagically" detecting everything about the code.
That said we do contemplate on adding import/export functionality, and if you feel strongly about it, please open an issue. If we have enough people requesting it, it will get done :)

This is interesting, but we can live without that at the moment. I will let anyone interested in that open an issue for that.

Now regarding your second point:

> -  `close() should wait for the analysis instead of giving it fixed 10sec grace.`

I would add a `Task.wait_for_detection()` that would be blocking until the task finished the detection. Then call that function inside `Task.close()`

> - argparser should somehow be called, otherwise, the argparser will not appear in the hyper-parameters section

I have no strong opinion here as I don't know how you read values from argparser internally. Can you infer them from the code or should the user pass them using `Task.connect(argparse.ArgumentParser())`? 

> - What happens if you do Task.current_task.connect({'somewhere_in_mid_code': 'inside_a_few_nested', 'functions': 1}) . Since this code is never reached when creating the "template" we will not be able to add these parameters into the "hyper-parameters" section

I would not worry about that. (Hyper) Parameters are what are known _before_ starting the experiment. Everything that is not known before and only accessible **during/after** an experiment is a result, not a parameter. Therefore users should not need to call Task.connect when experiment started (unless they use it for a second task, for piping a result. In that case, no problem).

Regarding Task Context:

> - When you leave the context, will it exit the code automatically (i.e. call exit() ) ?

Since context manager are custom try/finally blocks, you can definitely call the task.exit() in the finally block

> - You would still need to reset the Task before enqueuing it, or do you imagine that when you leave the context, the task will end up in "draft" mode?

So I would simply check if `remote` is True when creating the context object, if yes, do not execute the user code, (inside the block), simply do the conversion to draft/enqueueing logic. 

After the context block, the task can only be:
-  enqueued if remote was true
- done/failed if remote was false

> - It seems the context itself should have a queue argument so that it knows where to enqueue the Task, correct?

Yes, this seems reasonable 👍

**Comment by bmartinn** [comment]
Hi @H4dr1en 

> We simply set sdk.development.vcs_repo_detect_async = false to make sure that the detection is done sequentially. Setting a sleep is less robust I think.

I totally forgot about that feature, :heart:  the hack!!!
Anyhow I think task.close() should actually wait for the detection to finish...

Regrading the context, I really like the idea, and the guys here added another use case, 
where they run the training for a few iterations (say on a CPU machine), then when they see nothing crashed, they want to automatically stop it and launch it on one of the remote machines with the *trains-agent* .

Here is what we came up with
```python
task = Task.init(...)

# some code here

with task.run_once_locally(queue='workers', exit='true', args=None):
    print('we are done, leaving the rest for the worker, good bye')

# real training code here
```
- `queue` is the queue name to enqueue the running task into, it also implies exit=True. If it's None, it will not queue the task.
- `exit` if True exit the process at the end of the context.
- `args` if the argparser was not called yet, this gives you the ability to pass the argparser, so that trains will call it internally and log all the parameters, we might add `parameters` as a dict of parameters, basically with doing the same thing.

What do you think?

**Comment by H4dr1en** [comment]
Happy to ready that the team liked the idea!

The `run_once_locally` signature is not very straightfoward IMO, especially the `exit` parameter:

```
task = Task.init(...)
with task.run_once_locally(queue='workers', exit=True,  args=None):
    # run locally
# run remotely
```
This code is equivalent to:
```
task = Task.init(...)
# run locally
```
Since `exit=True` will exit the process at the end of the context. Or am I missing something?
This is also not very natural: one would expect the code under the context to be run remotely, not locally. Also it does not feel natural to pass the `queue` and `args` to a context that runs... locally.

I would rather have it as:

```
task = Task.init(...)

# local code

with task.run_in_agent(queue="default", args=None):
    # training code
```
This seems more natural for me, WDYT?

**Comment by bmartinn** [comment]
@H4dr1en I see your point...
You are looking at the context as "run me in a remote machine" where as I look at it as "run me on my machine". 

I can see why putting the remote execution code in a special context makes sense when you run the code on your machine (quite trivial to "exclude" the part that will be executed somewhere else), but I fear that actually most of our code needs to run in the remote machine.

After all we are thinking about a long process of training, and I don't want to have to add a context over most of my code, which is the training code. I think it makes more sense to put my tiny set up code in its own context. That said, I can diffidently see the awkwardness of the `exit=True` flag which makes the code execution non-linear, which I think we can all agree is not preferable. 

WDYT ?

**Comment by bmartinn** [comment]
@achigeor & @H4dr1en v0.15 is out with (I think) the solution you were looking for :smile: 
See details [here](https://github.com/allegroai/trains/issues/138#issuecomment-637566867)

**Comment by H4dr1en** [comment]
This is very cool, thanks for the great work!
